fhir-r4/src/main/java/org/dhis2/fhir/adapter/fhir/transform/dhis/impl/aggregate/r4/R4FhirMeasureReportToDataValueSetTransformer.java

```diff
@@ -0,0 +1,208 @@
+import org.hl7.fhir.r4.model.MeasureReport;
+import org.springframework.stereotype.Component;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.math.BigDecimal;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * R4 specific version of FHIR MeasureReport to DHIS2 DataValueSet transformer.
+ *
+ * @author David Katuscak
+ */
+
+@Component
+public class R4FhirMeasureReportToDataValueSetTransformer extends AbstractFhirMeasureReportToDataValueSetTranformer
+{
+    private final ZoneId zoneId = ZoneId.systemDefault();
+
+    public R4FhirMeasureReportToDataValueSetTransformer( @Nonnull final ScriptExecutor scriptExecutor,
+        @Nonnull final TrackedEntityMetadataService trackedEntityMetadataService,
+        @Nonnull final OrganizationUnitService organizationUnitService,
+        @Nonnull final TrackedEntityService trackedEntityService,
+        @Nonnull final FhirDhisAssignmentRepository fhirDhisAssignmentRepository,
+        @Nonnull ScriptExecutionContext scriptExecutionContext,
+        @Nonnull ValueConverter valueConverter )
+    {
+        super( scriptExecutor, trackedEntityMetadataService, organizationUnitService,
+            trackedEntityService, fhirDhisAssignmentRepository, scriptExecutionContext,
+            valueConverter );
+    }
+
+    @Nonnull
+    @Override
+    public Set<FhirVersion> getFhirVersions()
+    {
+        return FhirVersion.R4_ONLY;
+    }
+
+    @Override
+    public FhirToDhisTransformOutcome<DataValueSet> transformInternal( @Nonnull final FhirClientResource fhirClientResource,
+        @Nonnull final FhirToDhisTransformerContext context, @Nonnull final IBaseResource input,
+        @Nonnull final RuleInfo<DataValueSetRule> ruleInfo, @Nonnull final Map<String, Object> scriptVariables,
+        @Nonnull final DataValueSet dataValueSet )
+        throws TransformerException
+    {
+        final Map<String, Object> variables = new HashMap<>( scriptVariables );
+        final MeasureReport mr = (MeasureReport) input;
+
+        if ( ruleInfo.getRule().getDataSetIdLookupScript() == null )
+        {
+            logger.info( "Rule does not define a data value set lookup script and data value set does not yet include one." );
+            return null;
+        }
+        else
+        {
+            Optional<String> dataSetId = getDataSetId( context, ruleInfo, ruleInfo.getRule().getDataSetIdLookupScript(), variables );
+            dataSetId.ifPresent( dataValueSet::setDataSetId );
+        }
+
+        if ( dataValueSet.getDataSetId() == null )
+        {
+            throw new TransformerDataException( "Data Value Set ID cannot be decided." );
+        }
+
+        try
+        {
+            if ( Objects.isNull( mr.getPeriod() ) || Objects.isNull( mr.getPeriod().getStart() ) || Objects.isNull( mr.getPeriod().getEnd() ) )
```

Comment:
getPeriod() should never be null (contract of HAPI FHIR API, objects and lists and created always, isEmpty() can be used for null check).